gcs/constraints/regular.cc

```diff
@@ -0,0 +1,304 @@
+                            proof_step << "u ";
+                            state.for_each_guess([&](const Literal &lit) {
+                                if (!is_literally_true(lit))
+                                    proof_step << " 1 " << proof.proof_variable(!lit);
+                            });
+                            proof_step << " 1 " << proof.proof_variable(! state_at_pos_flags[i][q]);
+                            proof_step << " 1 " << proof.proof_variable(! state_at_pos_flags[i+1][next_q]);
+                            proof_step << " >= 1 ;\n";
+                            proof.emit_proof_line(proof_step.str());
+                        }
+                        });
+                    }
+
+                    // Finally, can eliminate what want
+                    state.infer(TrueLiteral{}, JustifyExplicitly{[&](Proof & proof, vector<ProofLine> &) -> void {
+                        stringstream proof_step;
+                        proof_step << "u ";
+                        state.for_each_guess([&](const Literal &lit) {
+                            if (!is_literally_true(lit))
+                                proof_step << " 1 " << proof.proof_variable(!lit);
+                        });
+                        proof_step << " 1 " << proof.proof_variable(! state_at_pos_flags[i+1][next_q]);
+                        proof_step << " >= 1 ;\n";
+                        proof.emit_proof_line(proof_step.str());
+                    }
+                    });
+                }
+            }
+        }
+        set<long> possible_final_states;
+        for (const auto & f : final_states) {
+            if (graph_nodes[num_vars].contains(f))
+                possible_final_states.insert(f);
+        }
+
+        graph_nodes[num_vars] = possible_final_states;
+
+        // Backward phase: validate
+        for (long i = num_vars - 1; i >= 0; --i) {
+
+            unordered_map<long, bool> state_is_support;
+            for (const auto & q : graph_nodes[i]) {
+                state_is_support[q] = false;
+            }
+
+            state.for_each_value(vars[i], [&](Integer val) -> void {
+                set<long> states = states_supporting[i][val];
+                for (const auto & q : states) {
+
+                    if (graph_nodes[i + 1].contains(transitions[q][val.raw_value])) {
+
+                        out_edges[i][q].emplace_back(transitions[q][val.raw_value]);
+                        out_deg[i][q]++;
+                        in_edges[i + 1][transitions[q][val.raw_value]].emplace_back(q);
+                        in_deg[i][transitions[q][val.raw_value]]++;
+                        state_is_support[q] = true;
+                    }
+                    else {
+
+                        states_supporting[i][val].erase(q);
+                        state.infer(TrueLiteral{}, JustifyExplicitly{[&](Proof & proof, vector<ProofLine> &) -> void {
+                            stringstream proof_step;
```

Comment:
There's a `Proof::trail_variables` that should cut this boilerplate down slightly. Maybe it's not the best name for it.